### PR TITLE
Local styles plugin

### PR DESCRIPTION
**Pull request by strarsis**
This PR adds the local styles plugin.
It is a complete rewrite of previous PR, to make the code simpler and more efficient.

With best regards

### CODE REVIEW

**Comment by strarsis** [comment]
Hm, jshint fails with build using nodejs release v0.10.40.
Should I base off from a different commit than current master?

With best regards

File: plugins/localStyles.js [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+     * Recursively traverses the HTML AST
+     *
+     * @param {Array}    items input items
+     * @param {Function} fn for processing an item
+     * @param {Function} fn for selecting an item
+     * @param {Function} fn for recursion check
+     *
+     * @return {Array} output items
+     */
+    function monkeys(items, callFn, selectFn, recurseFn) {
```
Comment (GreLI) [inline plugins/localStyles.js]: It seems that `recurseFn` isn't actually used.

File: plugins/localStyles.js [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+        while(i < length) {
+
+            var item = items.content[i];
+
+            if ( selectFn(item) ) {
+              callFn(item);
+            }
+
+            // recurse
+            if(item.content && item.content.length > 0 &&
```
Comment (GreLI) [inline plugins/localStyles.js]: You can use 'isEmpty()' method of API. See [lib/svgo/jsAPI.js#L87](https://github.com/svg/svgo/blob/master/lib/svgo/jsAPI.js#L87).

File: plugins/localStyles.js [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+      // Additionally parse inline CSS from the element (if there is any inline CSS)
+      var inlineAstRule = { declarations: [] };
+      if(elemHasStyles(matchedElement)) {
+        inlineAstRule = getInlineCss(matchedElement).stylesheet.rules[0];
+      }
+
+      // Merge inline CSS and looked up CSS
+      var newAstDeclarations =
+                 styleAstRule.declarations   // inline styles last for precedence
+        .concat(inlineAstRule.declarations);
```
Comment (GreLI) [inline plugins/localStyles.js]: Jumping indentation.

File: plugins/localStyles.js [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+
+      // Store the new (inlined + existing inline) CSS
+      // We don't stringify at match because this would be unnecessarily costly,
+      // when there are more selectors that match.
+      if(!matchedElement.newAstDeclarations)  matchedElement.newAstDeclarations = [];
+          matchedElement.newAstDeclarations = matchedElement.newAstDeclarations.concat(newAstDeclarations);
```
Comment (GreLI) [inline plugins/localStyles.js]: Line is indented as it's under `if` condition but it's not.

**Comment by GreLI** [comment]
Also, I'd like if styles were transfered just if there is only one match, not if two or more (non-adjanced, adjanced can be grouped). It seems it can be done easy enough in the current implementation.

**Comment by strarsis** [comment]
"Also, I'd like if styles were transfered just if there is only one match, not if two or more (non-adjanced, adjanced can be grouped). It seems it can be done easy enough in the current implementation."

So this means that when a CSS rule with multiple selectors 
is selected because _one_ selector matched /not all of them), 
then the whole CSS rule should be cleaned?

Like for example using this rule in input:
````
.test1, test2 {
    color: blue;
}
.test3 {
    background: red;
}
[...]
<rect class="test1" [...]
````
The rule is cleaned up completely because .test1 matched?:
````
.test3 {
    background: red;
}
[...]
<rect class="test1" [...]
````

With best regards

**Comment by strarsis** [comment]
The jshint fails with other files for this PR, should I base it off a different commit that passes the tests?
https://travis-ci.org/svg/svgo/jobs/91121264#L176

With best regards

**Comment by GreLI** [comment]
No, I don't think it's required. Do it if you want.

It's odd that other builds including master branch passed it.

**Comment by strarsis** [comment]
OK, so the last thing that needs to be clarified is how matched CSS rules are handled.
So a CSS rule should be cleaned up when any of its selectors matched, right?

With best regards

**Comment by GreLI** [comment]
I mean remove a rule only if there only one corresponding class in document. E.g.
```xml
<svg>
<style>
  .thick { stroke-width: 10px }
</style>
<path class="thick" d="..."/>
</svg>
```
but not then there are several classes here and there:
```xml
<svg>
<style>
  .thick { stroke-width: 10px }
</style>
...
<circle class="thick" .../>
...
<path class="thick" d="..."/>
...etc...
</svg>
```

File: plugins/localStyles.js [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+     *
+     * @return {Array} output items
+     */
+    function monkeys(items, callFn, selectFn) {
+      if(selectFn  === undefined) { selectFn  = selectAll;     }
+
+        var      i = 0,
+            length = items.content.length;
+
+        while(i < length) {
```
Comment (GreLI) [inline plugins/localStyles.js]: Why not just simple `for` loop?
```js
for (var i = 0; i < items.content.length; i++) {
```
Modern engines optimizes it, so there is no need in caching `length = items.content.length`.

**Comment by strarsis** [comment]
Yes, I changed it.
I also try to get rid of the global variables and add some further tests for edge cases.

With best regards

**Comment by GreLI** [comment]
Don't forget to check with «CDATA», empty `<style/>` tag, complex selectors, etc. Just fixed a couple of issues: fc174aa9f0412061a427ab592b7784f63c6ed170.

**Comment by LeZuse** [comment]
Do you guys have any update on this?

**Comment by strarsis** [comment]
Yes, I rewrote the plugin so that it can use the svgo ast as cheerio ast and juice for cleaning up the styles.
cheerio ast back to svgo ast is not fully working yet, hence no new PR has been created yet.

**Comment by strarsis** [comment]
OK, the rewrite has been finished: https://github.com/svg/svgo/pull/592
